chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use common::{
+    chain::{
+        timelock::OutputTimeLock, AccountSpending, ChainConfig, Transaction, TxInput, TxOutput,
+    },
+    primitives::{Amount, BlockDistance, BlockHeight},
+};
+use pos_accounting::PoSAccountingView;
+
+use crate::error::ConnectTransactionError;
+
+use super::IOPolicyError;
+
+/// `ConstrainedValueAccumulator` helps avoiding messy inputs/outputs combinations analysis by
+/// providing a set of properties that should be satisfied. For example instead of checking that
+/// all outputs are timelocked when the pool is decommissioned `ConstrainedValueAccumulator` gives a way
+/// to check that an accumulated output value is locked for sufficient amount of time which allows
+/// using other valid inputs and outputs in the same tx.
+///
+/// TODO: potentially this struct can be extended to collect unconstrained values such as transferred amounts
+///       replacing `AmountsMap`
+pub struct ConstrainedValueAccumulator {
+    timelock_requirement: BTreeMap<BlockDistance, Amount>,
+}
+
+impl ConstrainedValueAccumulator {
+    pub fn new() -> Self {
+        Self {
+            timelock_requirement: Default::default(),
+        }
+    }
+
+    pub fn collect_and_verify(
+        &mut self,
+        tx: &Transaction,
+        chain_config: &ChainConfig,
+        block_height: BlockHeight,
+        pos_accounting_view: &impl PoSAccountingView,
+        utxo_view: &impl utxo::UtxosView,
```

Comment:
Like we discussed, this should be simplified such that it doesn't have information about the utxo set or the account state. These should be pulled beforehand and put in the function parameters, so that we can have security proofs about the ConstrainedValueAccumulator.